### ISSUE TITLE
Cannot parse refinements definition on a type class 

### ISSUE THREAD

**Issue by kubum** [ucsd-progsys/liquidhaskell#1481]
For example:

```haskell
{-# LANGUAGE MultiParamTypeClasses #-}

{-@ 
class Monad m => Filter filter m where
  q :: forall a . Int -> Int -> Int -> m (filter a)
  qq :: forall a . filter a -> m Int
@-}
class Monad m => Filter filter m where
  q :: Int -> Int -> Int -> m (filter a)
  qq :: filter a -> m Int
```

Results in the error

```bash
8 |   qq :: forall a . filter a -> m Int
          ^

     unexpected ':'
     expecting bareTyArgP, "->", "=>", "(", white space or end of input
```

If we delete `qq`, liquidhaskell 0.8.2.4 parses it all fine.

According to the Basics.lhs example the functionality is supported [1], but it is tricky to see why the parser is unhappy.

[1] https://github.com/ucsd-progsys/liquidhaskell/blob/develop/docs/slides/HS2014/Basics.lhs#L209-L217

**Comment by ranjitjhala** [comment]
Can you remove the `Monad m => ` constraint from the _refined_ class definition?

**Comment by ranjitjhala** [comment]
no that's not it...

**Comment by ranjitjhala** [comment]
Hmm, not sure what's going on here - however, 
note that your spec doesn't actually put any 
refinements on the methods, so as such you 
can omit it?

However, I did try to put some refinements in 
which case it crashes LH. Thanks for posting 
this, we'll take a look.

(For now, I recommend working around this by
pulling the methods out as plain functions and
putting the refinements on them...)

**Comment by kubum** [comment]
Hi @ranjitjhala! Thank you for the response, yeah I found in some test cases there was used `;` to satisfy parser, but it didn't work in this case.

How do you see the extraction? I did it for the instances, a boring example might look as follows:

```haskell
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FlexibleInstances #-}

class Monad m => Filter filter m where
  q :: Int -> Int -> Int -> m (filter a)

data Table a = A | B

{-@ type PositiveInt = { v: Int | v > 0 } @-}

instance Monad m => Filter Table m where
  {-@
  instance Monad m => Filter Table m where
    q :: PositiveInt -> PositiveInt -> PositiveInt -> m (Table a)
  @-}
  q n1 n2 n3 = pure $ A
```

Result:

```bash
 13 | instance Monad m => Filter Table m where
               ^^^^^^^^^^^^^^^^^^^^^^^^^

   Inferred type
     VV : Int

   not a subtype of Required type
     VV : {VV : Int | VV > 0}

   In Context
```

Is it expected behaviour? The context is empty, and it looks like nothing has been violated yet, but maybe I am missing something.

**Comment by ranjitjhala** [comment]
Hi @kubum  -- sorry for the delay! I meant to pull the refinement outside the instance itself. It looks like the code for that is not working, and regardless, we've come to the conclusion that it can be unsound to 
have refinements on _instance_ that are different from those in the actual _class_ -- it can produce some quite counterintuitive effects! So I meant something like this, where the refinement is on a plain function
that is implemented by calling into the method.

http://goto.ucsd.edu:8090/index.html#?demo=permalink%2F1557372758_22294.hs

```haskell
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FlexibleInstances #-}


class Monad m => Filter filter m where
  q :: Int -> Int -> Int -> m (filter a)

instance Monad m => Filter Table m where
  q n1 n2 n3 = pure A
  
data Table a = A | B

{-@ type PositiveInt = { v: Int | v > 0 } @-}

{-@ qT :: PositiveInt -> PositiveInt -> PositiveInt -> m (Table a) @-}
qT :: (Monad m) => Int -> Int -> Int -> m (Table a)
qT = q 

ok :: IO (Table a)
ok  = qT 1 2 3            -- this is safe

bad :: IO (Table a)
bad = qT 0 2 3            -- this is rejected
```

**Comment by kosmikus** [comment]
Fixed via #1704.
